# ¡Hola Efren! <a class="tocSkip"></a>

Hola! Mi nombre es Nicolás Carrazco y ¡seré el revisor de tu proyecto! ¡Qué emoción!

Para esta primera vuelta, mi rol será como el de un buen jefe en tu futuro trabajo: voy a señalarte dónde hay algo por revisar, pero serás tú quien investigue y descubra el qué y el cómo solucionarlo. ¡Confío plenamente en que podrás con ello! Esto te prepara para el mundo real, donde un líder te guía pero no hace el trabajo por ti. Si después de intentarlo aún se te traba, no te preocupes, en la siguiente iteración te daré pistas más directas. ¡El objetivo es que aprendas un montón y que logres cosas increíbles!

Te voy a dejar mis comentarios justo aquí abajo. Por favor, no los muevas ni los modifiques, así nos aseguramos de que no se pierde nada.

Comenzaré haciéndote un resumen de tus puntos fuertes, aquellos que necesitan una corrección y aquellos que puedes llevar al siguiente nivel. Luego, te toca revisar todo el notebook para encontrar mis comentarios detallados, que estarán en cuadritos de colores (¡como semáforo!): verde, amarillo o rojo, que significan lo siguiente:

<div class='alert alert-block alert-success'>
<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Muy bien! Toda la respuesta fue lograda satisfactoriamente.
</div>

<div class='alert alert-block alert-warning'>
<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Existen detalles a mejorar. Existen recomendaciones.
</div>

<div class='alert alert-block alert-danger'>

<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Se necesitan correcciones en el bloque. El trabajo no puede ser aceptado con comentarios en rojo sin solucionar.
</div>

Cualquier comentario que quieras agregar entre iteraciones de revisión lo puedes hacer de la siguiente manera:

<div class='alert alert-block alert-info'>
<b>Respuesta estudiante.</b> <a class='tocSkip'></a>
</div>

¡Éxitos!

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General

Hola Efren, espero que te encuentres muy bien.  

Hiciste un proyecto perfecto!

Es por esto que voy a **aprobar tu proyecto**
</div>


# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
import pandas as pd # Importa pandas


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Bien!

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
df = pd.read_csv('/datasets/music_project_en.csv') # Lee el archivo y almacénalo en df


<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
print(df.head(10)) # Obtén las 10 primeras filas de la tabla df


     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
print(df.info()) # Obtén la información general sobre nuestros datos


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

¡Excelente!
   
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Algunos encabezados tienen espacios al principio.




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Todos los datos son del tipo: object. La forma correcta de conocer realmente que almacenan las columnas, sería pidiendo la información detallada del DataFrame al creador de la misma

2. Siento que tenemos los datos suficientes para dar respuesta a nuestra hipótesis

3. En las columnas 'Track', 'artist' y 'genre' hay datos ausentes. Aún no conocemos si hay datos duplicados. 

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Perfecto, buen trabajo desarrollando las observaciones. 
</div>


## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
print(df.columns) # Muestra los nombres de las columnas


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

¡Bien!
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
columnas_min = []
for columnas in df.columns:  # Bucle que itera sobre los encabezados y los pone todos en minúsculas
    a_minusculas = columnas.lower()
    columnas_min.append(a_minusculas)
    
df.columns = columnas_min
print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Buen trabajo usando el bucle de for para modificar los encabezados con y `lower()`.
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [7]:
new_col_names = []
for columnas in df.columns:  # Bucle que itera sobre los encabezados y elimina los espacios
    sin_espacios = columnas.strip()
    new_col_names.append(sin_espacios)
    
df.columns = new_col_names
print(df.columns)


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
df.rename(columns={'userid':'user_id'}, inplace=True) # Cambia el nombre de la columna "userid"


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
print(df.columns) # Comprueba el resultado: lista de encabezados


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

¡Excelente! Buen trabajo aplicando el formato <strong>snake_case</strong> a la variable <code>user_id</code> y estableciendo como encabezados del dataset la lista con las modificaciones realizadas. 
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [10]:
print(df.isna().sum()) # Calcula el número de valores ausentes


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [11]:
columns_to_replace = ['track', 'artist', 'genre']

for col in columns_to_replace: # Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
	df[col] = df[col].fillna('unknown') 


<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [12]:
print(df.isna().sum()) # Cuenta los valores ausentes


user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Excelente, buen trabajo arreglando los valores nulos.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [13]:
print(df.duplicated().sum()) # Cuenta los duplicados explícitos

3826


<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
df = df.drop_duplicates().reset_index(drop=True) # Elimina los duplicados explícitos

<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
print(df.duplicated().sum()) # Comprueba de nuevo si hay duplicados

0


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el método <code>drop_duplicates()</code> para solucionar el problema de los valores duplicados en el dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
generos_unicos = df['genre'].unique()
ordenados = sorted(generos_unicos)
print(ordenados)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [17]:
def replace_wrong_values(df, column, wrong_values, correct_value): # Función para reemplazar los duplicados implícitos
    for wrong_value in wrong_values: # un bucle sobre nombres mal escritos
        df[column] = df[column].replace(wrong_value, correct_value) # llamar a replace() para cada nombre incorrecto
    return df 


<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [18]:
duplicates = ['hip', 'hop', 'hip-hop'] # una lista de nombres mal escritos
name = 'hiphop' # el nombre correcto
df = replace_wrong_values(df, 'genre', duplicates, name) # Elimina los duplicados implícitos


<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
generos_unicos = df['genre'].unique()
ordenados = sorted(generos_unicos)
print(ordenados) # Comprueba de nuevo los duplicados implícitos


['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [20]:
lunes = df[df["day"] == "Monday"]
lunes_por_ciudad = lunes.groupby('city')['track'].count() # Cuenta las canciones reproducidas en cada ciudad

print('\nLas canciones escuchadas los LUNES son:\n', lunes_por_ciudad)

viernes = df[df["day"] == "Friday"]
viernes_por_ciudad = viernes.groupby('city')['track'].count() # Cuenta las canciones reproducidas en cada ciudad

print('\nLas canciones escuchadas los VIERNES son:\n',viernes_por_ciudad)


Las canciones escuchadas los LUNES son:
 city
Shelbyville     5614
Springfield    15740
Name: track, dtype: int64

Las canciones escuchadas los VIERNES son:
 city
Shelbyville     5895
Springfield    15945
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?

1. En ambas ciudades, se reproducen más canciones los viernes, ¿comienza el fin de semana?
2. En ambos días, en Springfield se reproducen más canciones que en Shelbyville, más del triple en ambos días. Springfield tiene más habitantes

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Bien, buena observación.

</div>


Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [21]:
por_dia = df.groupby('day')['track'].count() # Calcula las canciones reproducidas en cada uno de los dos días

print(por_dia)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

El viernes tiene mas canciones reproducidas qu el lunes, lo que también se obtuvo de analizar los días por ciudad.

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [22]:
def number_tracks(df, day, city): # Declara la función number_tracks() con dos parámetros: day= y city=.
    dia = df[df['day'] == day] # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    ciudad = dia[dia['city'] == city] # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    usuarios = ciudad['user_id'].count() # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    return(usuarios) # Devuelve el número de valores de la columna 'user_id'

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Excelente, definiste bien la función `number_tracks` ingresando a los valores de day y city de manera excelente.
    
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [23]:
print(number_tracks(df, 'Monday', 'Springfield'))  # El número de canciones reproducidas en Springfield el lunes

15740


In [24]:
print(number_tracks(df, 'Monday', 'Shelbyville')) # El número de canciones reproducidas en Shelbyville el lunes

5614


In [25]:
print(number_tracks(df, 'Friday', 'Springfield')) # El número de canciones reproducidas en Springfield el viernes

15945


In [26]:
print(number_tracks(df, 'Friday', 'Shelbyville')) # El número de canciones reproducidas en Shelbyville el viernes

5895


<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Muy buen bloque de análisis.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


En general, se observó que las reproducciones varían de acuerdo a la ciudad y el día de la semana. En Springfield se reproducen más canciones sin importar el día de la semana, mientras que los viernes se reproducen más canciones sin importan la ciudad, lo que observamos en la tabla:

city           lunes   viernes    total
Shelbyville     5614      5895    11509
Springfield    15740     15945    31685
        total  21354     21840

Los principales problemas encontrados fueron:

1. Los estilos de los encabezados, se trato con la regla snake_case
2. Valores ausentes en algunas columnas, se sustituyeron por el string 'unknown'
3. Filas idénticas (explícitos), se eliminaron
4. Nombres mal escritos en la columna 'genre', se homologaron

Todo lo anterior permitió que el análisis fuera más confiable utilizando los filtros necesarios.

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

¡Totalmente perfecto!

Tus conclusiones son muy acertadas y basadas en datos reales, me gusta cómo explicas también, sigue así.
</div>
